src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorScriptDocValues.java

```diff
@@ -0,0 +1,72 @@
+/*
+ *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.knn.index;
+
+import org.apache.lucene.index.BinaryDocValues;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.index.fielddata.ScriptDocValues;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+
+// This class is thread safe, since docExists is synchronized at an instance level
+public final class KNNVectorScriptDocValues extends ScriptDocValues<float[]> {
```

Comment:
Should this extend BinaryScriptDocValues<float[]>? Here is the [code](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/index/fielddata/ScriptDocValues.java#L440).

I think if we go this route, we do not have to have docExists boolean.

Reply:
Thats a good point. That was my first attempt, but BinaryScriptDocValues is not public, hence we can't extend it.

Reply:
Thats a good point. That was my first attempt, but BinaryScriptDocValues is not public, hence we can't extend it.

Reply:
Ah, I see. That makes sense. In terms of docExists, I think we should follow the convention of using [count](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/index/fielddata/ScriptDocValues.java#L493) instead. What do you think?

Reply:
We can do that too, but i kept consistent with our other implementation as here
https://github.com/opendistro-for-elasticsearch/k-NN/blob/master/src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/script/KNNScoreScript.java#L116

Reply:
You are right. We probably don't need to make that change now.